Realm/Realm/Handles/SharedRealmHandle.cs

```diff
@@ -767,6 +774,13 @@ private static void HandleTaskCompletionCallback(IntPtr tcs_ptr, NativeException
 
         private static void HandleTaskCompletion<T>(IntPtr tcsPtr, Func<T> resultBuilder, NativeException ex)
         {
+            // if the cancellation of the task is executed just an instant before core calls this callback,
+            // the handle is already freed by the finally block of the managed caller.
+            if (tcsPtr == IntPtr.Zero)
+            {
+                return;
+            }
```

Comment:
I don't think `tcsPtr` can ever be zero here. Even if the handle is freed by the finally block, we would have already passed a non-zero value to Core, which we'd get here.

That being said, we're in a bit of a trouble here, because Mono has a bug where calling `GCHandle.FromIntPtr` on a freed gchandle will crash the process. So if there's a chance that this callback gets invoked after we free the handle, we'll need to rework the code to make sure it never happens.

Reply:
> I don't think tcsPtr can ever be zero here. Even if the handle is freed by the finally block, we would have already passed a non-zero value to Core, which we'd get here.

You're right. I misjudged that. I was trying to avoid managed to resolve the handle when the pointer points to garbage.

> we're in a bit of a trouble here, because Mono has a bug where calling GCHandle.FromIntPtr on a freed gchandle will crash the process

Not good

> if there's a chance that this callback gets invoked after we free the handle

I have not seen this happening while testing, but I thought that being in a multi-threaded scenario we can't know what timing does, so I wanted to be on the safer side.

Reply:
So I can report that with some "good timing" core ends up calling our managed callback when we have already freed the `tcs`. A typical scenario for this would be the following:
1. `BeginWriteAsync` is called and the await starts
2. The managed cb is queued by core for later execution
3. The `CancellationToken` is called right when core is about to dequeue
4. At this point the finally block of `BeginTransactionAsync` and the `tcs` handle is freed
5. Core calls the managed cb (`SharedRealmHandle.HandleTaskCompletion`) and `GCHandle.FromIntPtr(tcsPtr);` returns a null reference.

Reply:
Then I'll move ownership of the `handle` to `HandleTaskCompletion` and add a bunch of comments around who owns what.